### PR TITLE
WIP: Object resolver interpreter

### PR DESCRIPTION
**Pull request by dkarlovi**
Allows the import process to resolve previously imported objects by an additional property.

For example, with a structure like this:

```
id pid name
--------------------
1  n/a Root
2    1 Category X
3    1 Category Y
4    2 Category X X
```

and a class definition `Category(originalId)`, we should be able to replicate the original tree structure by setting `o_parentId = object_resolver(field=pid, class=Category, column=originalId)`.

### CODE REVIEW

File: src/ImportDefinitionsBundle/Form/Type/Interpreter/ObjectResolverType.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\FormBuilderInterface;
+
+/**
+ * Class ObjectResolverType.
+ */
+class ObjectResolverType extends AbstractType
+{
+    /**
+     * @inheritdoc
```
Comment (dpfaffenbauer) [inline src/ImportDefinitionsBundle/Form/Type/Interpreter/ObjectResolverType.php]: use {@inheritdoc}  please

File: src/ImportDefinitionsBundle/Interpreter/ObjectResolver.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * @license    https://github.com/w-vision/ImportDefinitions/blob/master/gpl-3.0.txt GNU General Public License version 3 (GPLv3)
+ */
+
+namespace ImportDefinitionsBundle\Interpreter;
+
+use ImportDefinitionsBundle\Model\DefinitionInterface;
+use ImportDefinitionsBundle\Model\Mapping;
+use Pimcore\Model\DataObject\Concrete;
+
+/**
```
Comment (dpfaffenbauer) [inline src/ImportDefinitionsBundle/Interpreter/ObjectResolver.php]: please don't use a class header

File: src/ImportDefinitionsBundle/Form/Type/Interpreter/ObjectResolverType.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+namespace ImportDefinitionsBundle\Form\Type\Interpreter;
+
+use ImportDefinitionsBundle\Form\Type\ClassChoiceType;
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\FormBuilderInterface;
+
+/**
```
Comment (dpfaffenbauer) [inline src/ImportDefinitionsBundle/Form/Type/Interpreter/ObjectResolverType.php]: please don't use a class header

File: src/ImportDefinitionsBundle/Interpreter/ObjectResolver.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * Class ObjectResolver.
+ */
+class ObjectResolver implements InterpreterInterface
+{
+    /**
+     * @inheritdoc
```
Comment (dpfaffenbauer) [inline src/ImportDefinitionsBundle/Interpreter/ObjectResolver.php]: use {@inheritdoc} please

File: src/ImportDefinitionsBundle/Resources/public/pimcore/js/interpreters/objectresolver.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+ * For the full copyright and license information, please view the LICENSE.md and gpl-3.0.txt
+ * files that are distributed with this source code.
+ *
+ * @copyright  Copyright (c) 2016-2018 w-vision AG (https://www.w-vision.ch)
+ * @license    https://github.com/w-vision/ImportDefinitions/blob/master/gpl-3.0.txt GNU General Public License version 3 (GPLv3)
+ */
+
+pimcore.registerNS('pimcore.plugin.importdefinitions.interpreters.objectresolver');
+
+pimcore.plugin.importdefinitions.interpreters.expression = Class.create(pimcore.plugin.importdefinitions.interpreters.abstract, {
```
Comment (dpfaffenbauer) [inline src/ImportDefinitionsBundle/Resources/public/pimcore/js/interpreters/objectresolver.js]: should be ```pimcore.plugin.importdefinitions.interpreters.object_resolver = ```

**Comment by dpfaffenbauer** [comment]
Shouldn't the class be called ObjectIdResolver? Since the o_parentId field requires a ID not an object. Otherwise we could just introduce o_parent to ImportDefinitions as well. I quite like the idea of ObjectResolver better than just ObjectIdResolver. On the other hand, using nested interpreters it should be possible to combine ObjectIdResolver with Href (which loads a object by id)

**Comment by dkarlovi** [comment]
> Shouldn't the class be called ObjectIdResolver?

The way I see it, there's nothing forcing us to use IDs, those will just be the most common use case. You might use SKU or some other identifier which you have in your own data store you're importing.

**Comment by dkarlovi** [comment]
>  It doesn't resolve a DataObjectinstance, it resolves its ID

Fair, we should then use a better name for the functionality to signify it **looks up the object by an arbitrary field** and then returns Pimcore's object ID.

**Comment by dpfaffenbauer** [comment]
> Fair, we should then use a better name for the functionality to signify it looks up the object by an arbitrary field and then returns Pimcore's object ID.

Thats right: But what If it returns the Pimcore Object and we introduce a new `parent` field in the definitions column panel? Usually in Pimcore you always work directly with instances of Objects rather than ID's of Objects.

**Comment by dkarlovi** [comment]
> But what If it returns the Pimcore Object and we introduce a new parent field in the definitions column panel? 

Makes sense, but then you don't get to reuse it with Href?

**Comment by dpfaffenbauer** [comment]
Yes, but don't need that really, do we? Having the Object gives you more possibilities than having the id

**Comment by dkarlovi** [comment]
OK, how do we proceed here?

**Comment by dpfaffenbauer** [comment]
I just added the o_parent field to the allowed mapping fields (https://github.com/w-vision/ImportDefinitions/commit/f3fc7a7cb17c240febb19bded37484996c26b20d). So lets stick with the Object instance variant.

**Comment by dkarlovi** [comment]
@dpfaffenbauer do you think we should rename the interpreter or is `ObjectResolver` OK?

**Comment by dpfaffenbauer** [comment]
`ObjectResolver` is fine

File: src/ImportDefinitionsBundle/Interpreter/ObjectResolver.php [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+         */
+        $listing = $class::$lookup($value);
+        $found = $listing->count();
+
+        if ($found < 1 || $found > 1) {
+            // too few or too many found
+            return null;
+        }
+
+        return $listing->current()->getId();
```
Comment (dpfaffenbauer) [inline src/ImportDefinitionsBundle/Interpreter/ObjectResolver.php]: You need to return the actual object here instead of the id.

**Pull request merged**